### PR TITLE
[SAH-70]: Implement User Onboarding 

### LINKED ISSUE #148: [SAH-70]: Implement User Onboarding
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Maiz27**

### CODE REVIEW

**Review by mde3 — COMMENTED** [review]
I see you've updated all the next version across the apps, will that have any impact

**Comment by Maiz27** [comment]
> I see you've updated all the next version across the apps, will that have any impact

@mde3 They might impact some minor parts of the code but it was needed since the different versions across all the apps and packages were causing unusual errors when interacting with each other

File: packages/features/auth/forms/onboarding/Preferences.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    const validatedInput = schema.parse(data);
+
+    updateStepFormData({
+      preference: validatedInput[role],
+    });
+    navigateToNextStep("summary");
+  };
+
+  const renderForm = () => {
+    switch (role) {
```
Comment (Emmanuel-Melon) [inline packages/features/auth/forms/onboarding/Preferences.tsx:46]: Using a switch statement is cool because we have a relatively small set of roles. We could probably refactor this to conditionally render content by setting the role in the Zustand store when we start adding more roles.

You could probably create an issue for this later.

File: packages/features/auth/forms/onboarding/RoleDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  });
+
+  const onSubmit: SubmitHandler<FormData> = async (data) => {
+    const validatedInput = schema.parse(data);
+
+    updateStepFormData(validatedInput);
+    navigateToNextStep("preferences");
+  };
+
+  const renderForm = () => {
```
Comment (Emmanuel-Melon) [inline packages/features/auth/forms/onboarding/RoleDetails.tsx:38]: This is fine but we could consider reusing this function later on.

File: packages/features/auth/forms/onboarding/Summary.tsx [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    onboardSupplier,
+    data: supplierResult,
+    loading: supplierLoading,
+    error: supplierError,
+  } = useOnboardSupplier();
+
+  const role = formData.role as "supplier" | "business";
+  // Create schema in two steps to avoid TS error with
+  // computed properties (role) on combined schema.
+  const baseSchema = z.object({
```
Comment (Emmanuel-Melon) [inline packages/features/auth/forms/onboarding/Summary.tsx:41]: This is neat!

File: packages/features/auth/forms/onboarding/Summary.tsx [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    values: formData as any,
+  });
+
+  const onSubmit: SubmitHandler<FormData> = async () => {
+    const validatedInput = combinedSchema.parse(formData);
+
+    if (!businessLoading && !supplierLoading) {
+      const { business, supplier, preference } = validatedInput;
+
+      switch (role) {
```
Comment (Emmanuel-Melon) [inline packages/features/auth/forms/onboarding/Summary.tsx:65]: Third occurrence!

File: packages/features/auth/forms/onboarding/constants.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  name: z.string().min(2, { message: "Must be more than 2 characters" }),
+  type: z.string(),
+  contactName: z.string().min(2, { message: "Must be more than 2 characters" }),
+  description: z
+    .string()
+    .min(10, { message: "Must be more than 10 characters" }),
+  contactEmail: z.string().email("Invalid email"),
+  phoneNumber: z.string().refine(
+    (phone) => {
+      const internationalFormat = /^\+\d{1,3}\d{6,14}$/;
```
Comment (Emmanuel-Melon) [inline packages/features/auth/forms/onboarding/constants.ts:29]: Let's extract this to a function because we need to use this in multiple places, i.e., mobile number verification, deposits/ withdrawals, and code/ password resets.

File: packages/features/auth/lib/auth.ts [hunk 6]
```diff
@@ -61,12 +63,27 @@ const initNextAuth = (): AuthOptions => {
...
+        const { hasCompletedOnboarding, role } = users_by_pk;
+
         return {
           ...token,
+          hasCompletedOnboarding,
           "https://hasura.io/jwt/claims": {
             "x-hasura-allowed-roles": ["user"],
             "x-hasura-default-role": "user",
-            "x-hasura-role": "user",
+            "x-hasura-role": role ?? "user",
```
Comment (Emmanuel-Melon) [inline packages/features/auth/lib/auth.ts:86]: Neat!

File: packages/lib/hooks/formStores/useOnBoardingFormStore.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+export const INITIAL_STEP = "user_details";
+
+export const ON_BOARDING_STEPS = [
+  "user_details",
+  "role_details",
+  "preferences",
+  "summary",
+] as const;
+
+// eslint-disable-next-line react-hooks/rules-of-hooks
```
Comment (Emmanuel-Melon) [inline packages/lib/hooks/formStores/useOnBoardingFormStore.tsx:46]: @Maiz27  we might have to do something about this later on. Should we create an issue to explore a possible refactor? 

I think this works fine but these warnings do exist for a reason and we might end up with implicit bugs. I can't think of any at the moment but it might be worth considering, what do you think?

File: packages/ui/components/Radio.tsx [hunk 8]
```diff
@@ -1,36 +1,75 @@
...
 
-type RadioProps = {
+type Option = {
+  value: string;
   label: string;
-  icon: any;
-  name: string;
-  register: any;
-  errors: any;
+  icon?: any;
```
Comment (Emmanuel-Melon) [inline packages/ui/components/Radio.tsx:9]: I think there's a type for icon

**Review by Emmanuel-Melon — APPROVED** [review]
Good work!

File: packages/features/auth/forms/onboarding/Preferences.tsx (hunk 1 above)
Comment (Maiz27) [reply packages/features/auth/forms/onboarding/Preferences.tsx:46]: Make sense, I used switch for the potential of new roles getting added. With it, adding new roles is much easier

File: packages/lib/hooks/formStores/useOnBoardingFormStore.tsx (hunk 7 above)
Comment (Maiz27) [reply packages/lib/hooks/formStores/useOnBoardingFormStore.tsx:46]: Good idea!

File: packages/ui/components/Navbar.tsx
Comment (mde3) [inline packages/ui/components/Navbar.tsx:1]: Solid work my guy, I've wanted to work on improving the navbar component but just wanted to wait & view the changes you made then track back to improving it 

Dw it's nothing major just few tweaks so i guess that could go in a separate pr

**Review by mde3 — APPROVED** [review]

File: packages/ui/components/Navbar.tsx
Comment (Maiz27) [reply packages/ui/components/Navbar.tsx:1]: No worries man, feel free to do it after this gets merged!
